xen/arch/arm/mm.c

```diff
@@ -1287,20 +1305,31 @@ long arch_memory_op(int op, XEN_GUEST_HANDLE_PARAM(void) arg)
         if ( d == NULL )
             return -ESRCH;
 
-        for ( i = 0; i < req.num_frames; i++ )
+        if ( likely (req.num_frames <= 32) )
         {
-            xen_pfn_t pa, ma;
+            /* more than 95% cases for GSX operation */
+            xen_pfn_t pma[req.num_frames];
 
-            if ( unlikely(copy_from_guest_offset(&pa, req.pa, i, 1)) )
+            if ( unlikely(process_p2m_lookup(d, pma, &req)) )
             {
                 rcu_unlock_domain(d);
                 return -EFAULT;
             }
+        }
+        else
+        {
+            xen_pfn_t *pma;
+
+            pma = xmalloc_array(xen_pfn_t, req.num_frames);
 
-            ma = pfn_to_paddr(mfn_x(p2m_lookup(d, _gfn(paddr_to_pfn(pa)), NULL)));
+            if ( !pma )
+            {
+                return -ENOMEM;
+            }
 
-            if ( unlikely(copy_to_guest_offset(req.ma, i, &ma, 1)) )
+            if ( unlikely(process_p2m_lookup(d, pma, &req)) )
             {
+                xfree(pma);
```

Comment:
Why you don't free pma if process_p2m_lookup returns 0?

Reply:
Good question. Must be provtikav. And that was the root of an additional performance gain seen with this patch.